README.md
[![Gem Version](https://badge.fury.io/rb/payout.svg)](https://badge.fury.io/rb/payout) [![Build Status](https://travis-ci.org/payout/payout-ruby.svg?branch=master)](https://travis-ci.org/payout/payout-ruby) [![Code Climate](https://codeclimate.com/github/payout/payout-ruby/badges/gpa.svg)](https://codeclimate.com/github/payout/payout-ruby) [![Test Coverage](https://codeclimate.com/github/payout/payout-ruby/badges/coverage.svg)](https://codeclimate.com/github/payout/payout-ruby/coverage)

# Payout Ruby SDK

The Payout Ruby SDK allows you to more quickly integrate with the [Payout](http://www.payout.com) API.

## Installation

The SDK is available as gem, which you can install like so:

```
gem install payout
```

Alternatively, you can add it to you `Gemfile`.

```ruby
gem 'payout', '~> 0.1.0'
```

And then run `bundle install`.

## Ruby Versions
The SDK has been tested on Linux with the following ruby versions:
 * 1.9.3
 * 2.0.0-p247
 * 2.1.6
 * 2.2.2
 * 2.3.1

## Configuration

```ruby
Payout.api_url =      ENV['PAYOUT_API_URL']           # Defaults to https://live.payout.com
Payout.api_token =    ENV['PAYOUT_API_TOKEN']         # You can find your credentials at
Payout.api_secret =   ENV['PAYOUT_API_SECRET']        # https://dashboard.payout.com/api
Payout.open_timeout = ENV['PAYOUT_OPEN_TIMEOUT'].to_i # Defaults to 30
Payout.read_timeout = ENV['PAYOUT_READ_TIMEOUT'].to_i # Defaults to 80
```

## Documentation

See full documentation [here](https://www.payout.com/page/ruby-library).


CHANGELOG.md
# 0.2.0
Adding support for bank accounts.

# 0.1.0
Initial release supporting the version 1 balance, card and payment endpoints.


spec/payout_spec.rb
RSpec.describe Payout do
  describe '::version' do
    subject { Payout.version }
    it { is_expected.to eq Payout::VERSION }
  end # ::version

  describe '::api_version' do
    subject { Payout.api_version }

    context 'with being set to a string' do
      it 'should raise ArgumentError' do
        expect { Payout.api_version = '1' }.to raise_error ArgumentError,
          'must be an integer'
      end
    end

    context 'with being set to negative integer' do
      it 'should raise ArgumentError' do
        expect { Payout.api_version = -1 }.to raise_error ArgumentError,
          'must be a positive integer'
      end
    end

    context 'with being set to unsupported version' do
      it 'should raise ArgumentError' do
        expect { Payout.api_version = 1337 }.to raise_error ArgumentError,
          'unsupported version'
      end
    end

    context 'with it being automatically set by accessing a constant' do
      # This should cause the `const_missing` method to auto include the default
      # version.
      before { Payout::Card }

      it 'should equal default version' do
        is_expected.to be Payout::DEFAULT_API_VERSION
      end
    end

    context 'with it already having been initialized' do
      # This test assumes that the contexts are being run in order.
      # The 'without it being set' context test will cause the API_VERSION
      # constant to be defined. Since it can't be undefined we can't fully
      # isolate these tests.

      it 'should raise VersionError' do
        expect { Payout.api_version = Payout::DEFAULT_API_VERSION }
          .to raise_error Payout::VersionError, 'cannot change version after '\
            'it has been initialized'
      end
    end
  end # ::api_version

  describe '::const_missing' do
    # Here we can only test const_missing *after* the version has been
    # initialized since it gets initialized in the '::api_version' tests above.

    context 'with undefined constant' do
      it 'should raise standard NameError' do
        expect { Payout::UNDEFINED_CONSTANT }.to raise_error NameError,
          'uninitialized constant Payout::UNDEFINED_CONSTANT'
      end
    end
  end # ::const_missing

  describe '::api_url' do
    subject { Payout.api_url }

    context 'without having set value' do
      it { is_expected.to eq 'https://live.payout.com' }
    end

    context 'with it having been set to nil' do
      before { Payout.api_url = nil }
      it { is_expected.to eq 'https://live.payout.com' }
    end

    context 'with it having been sent to sandbox url' do
      before { Payout.api_url = 'https://sandbox.payout.com' }
      it { is_expected.to eq 'https://sandbox.payout.com' }
    end

    context 'with trailing forward-slash in url' do
      before { Payout.api_url = 'https://sandbox.payout.com/' }

      it 'should remove trailing forward-slash' do
        is_expected.to eq 'https://sandbox.payout.com'
      end
    end
  end # ::api_url

  describe '::api_token' do
    subject { Payout.api_token }

    context 'without having set a value' do
      it 'should raise AuthenticationError' do
        expect { subject }.to raise_error Payout::AuthenticationError,
          'Payout.api_token must be defined'
      end
    end

    context 'with it being set to valid value' do
      before { Payout.api_token = 'validtoken' }
      it { is_expected.to eq 'validtoken' }
    end
  end # ::api_token

  describe '::api_secret' do
    subject { Payout.api_secret }

    context 'without having set a value' do
      it 'should raise AuthenticationError' do
        expect { subject }.to raise_error Payout::AuthenticationError,
          'Payout.api_secret must be defined'
      end
    end

    context 'with it being set to valid value' do
      before { Payout.api_secret = 'validtoken' }
      it { is_expected.to eq 'validtoken' }
    end
  end # ::api_secret

  describe '::open_timeout' do
    subject { Payout.open_timeout }

    context 'without having set a value' do
      it { is_expected.to eq Payout::DEFAULT_OPEN_TIMEOUT }
    end

    context 'with having been set to 15' do
      before { Payout.open_timeout = 15 }
      it { is_expected.to eq 15 }
    end

    context 'with being set to 15.0' do
      it 'should raise ArgumentError' do
        expect { Payout.open_timeout = 15.0 }.to raise_error ArgumentError,
          'must be an integer'
      end
    end

    context 'with being set to "seven"' do
      it 'should raise ArgumentError' do
        expect { Payout.open_timeout = 'seven' }.to raise_error ArgumentError,
          'must be an integer'
      end
    end
  end # ::open_timeout

  describe '::read_timeout' do
    subject { Payout.read_timeout }

    context 'without having set a value' do
      it { is_expected.to eq Payout::DEFAULT_READ_TIMEOUT }
    end

    context 'with having been set to 15' do
      before { Payout.read_timeout = 15 }
      it { is_expected.to eq 15 }
    end

    context 'with being set to 15.0' do
      it 'should raise ArgumentError' do
        expect { Payout.read_timeout = 15.0 }.to raise_error ArgumentError,
          'must be an integer'
      end
    end

    context 'with being set to "seven"' do
      it 'should raise ArgumentError' do
        expect { Payout.read_timeout = 'seven' }.to raise_error ArgumentError,
          'must be an integer'
      end
    end
  end # ::read_timeout

  describe '::request' do
    subject { Payout.request(verb, path, params) }
    let(:verb) { :get }
    let(:params) { {} }
    let(:path) { '/test_path' }

    before do
      Payout.api_token = api_token
      Payout.api_secret = api_secret
    end

    let(:api_token) { 'api_token' }
    let(:api_secret) { 'api_secret' }

    it 'should set verify_ssl to verify peer' do
      should_execute_with(verify_ssl: OpenSSL::SSL::VERIFY_PEER)
    end

    it 'should set ssl_ca_file to Payout::SSL_CA_FILE' do
      should_execute_with(ssl_ca_file: Payout::SSL_CA_FILE)
    end

    context 'with open_timeout unset' do
      it { should_execute_with(open_timeout: Payout::DEFAULT_OPEN_TIMEOUT) }
    end

    context 'with open_timeout = 11' do
      before { Payout.open_timeout = 11 }
      it { should_execute_with(open_timeout: 11) }
    end

    context 'with read_timeout unset' do
      it { should_execute_with(read_timeout: Payout::DEFAULT_READ_TIMEOUT) }
    end

    context 'with read_timeout = 11' do
      before { Payout.read_timeout = 11 }
      it { should_execute_with(read_timeout: 11) }
    end

    context 'with verb = :get' do
      let(:verb) { :get }
      it { should_execute_with(method: :get) }

      context 'with params = nil' do
        let(:params) { nil }

        it 'should not pass params in headers' do
          should_execute_with(headers: hash_excluding(:params))
        end
      end

      context 'with params = {}' do
        let(:params) { {} }

        it 'should not pass params in headers' do
          should_execute_with(headers: hash_excluding(:params))
        end
      end

      context 'with unempty params' do
        let(:params) { { a: 1, b: 2, c: 3 } }

        it 'should pass params in headers' do
          should_execute_with(headers: hash_including(params: params))
        end
      end
    end # with verb = :get

    context 'with verb = :delete' do
      let(:verb) { :delete }
      it { should_execute_with(method: :delete) }

      context 'with params = nil' do
        let(:params) { nil }

        it 'should not pass params in headers' do
          should_execute_with(headers: hash_excluding(:params))
        end
      end

      context 'with params = {}' do
        let(:params) { {} }

        it 'should not pass params in headers' do
          should_execute_with(headers: hash_excluding(:params))
        end
      end

      context 'with unempty params' do
        let(:params) { { a: 1, b: 2, c: 3 } }

        it 'should pass params in headers' do
          should_execute_with(headers: hash_including(params: params))
        end
      end
    end # with verb = :delete

    context 'with verb = :post' do
      let(:verb) { :post }
      it { should_execute_with(method: :post) }

      context 'with params = nil' do
        let(:params) { nil }

        it 'should pass nil payload' do
          should_execute_with(payload: nil)
        end
      end

      context 'with params = {}' do
        let(:params) { {} }

        it 'should pass nil payload' do
          should_execute_with(payload: nil)
        end
      end

      context 'with unempty params' do
        let(:params) { { a: 1, b: 2, c: 3 } }

        it 'should pass params as JSON payload' do
          should_execute_with(payload: params.to_json)
        end

        it 'should set Content-Type: application/json' do
          should_execute_with(
            headers: hash_including(content_type: 'application/json')
          )
        end
      end
    end # with verb = :post

    context 'with verb = :put' do
      let(:verb) { :put }
      it { should_execute_with(method: :put) }

      context 'with params = nil' do
        let(:params) { nil }

        it 'should pass nil payload' do
          should_execute_with(payload: nil)
        end
      end

      context 'with params = {}' do
        let(:params) { {} }

        it 'should pass nil payload' do
          should_execute_with(payload: nil)
        end
      end

      context 'with unempty params' do
        let(:params) { { a: 1, b: 2, c: 3 } }

        it 'should pass params as JSON payload' do
          should_execute_with(payload: params.to_json)
        end

        it 'should set Content-Type: application/json' do
          should_execute_with(
            headers: hash_including(content_type: 'application/json')
          )
        end
      end
    end # with verb = :put

    context 'with verb = :bad' do
      let(:verb) { :bad }

      it 'should raise ArgumentError' do
        expect { subject }.to raise_error ArgumentError,
          'invalid request verb: :bad'
      end
    end # with verb = :bad

    context 'with path without initial forward-slash' do
      let(:path) { 'test_path' }

      it 'should build URL correctly' do
        should_execute_with(url: "https://live.payout.com/#{path}")
      end
    end
  end # ::request
end # Payout


spec/spec_helper.rb
require "codeclimate-test-reporter"
CodeClimate::TestReporter.start

require 'payout'

# Monkey patch request logic
module Payout
  class << self
    def mock_client
      @mock_client
    end

    def within_test
      ##
      # Make sure all instance variables are nil to similate a fresh state.
      Payout.instance_variables.each do |v|
        Payout.instance_variable_set(v, nil)
      end

      @mock_client = RSpec::Mocks::Double.new('MockClient')
      yield
      @mock_client = nil
    end

    ##
    # Allows the ::request method to be mocked during the duration of the
    # current test.
    def mock_request
      @mock_request = true
    end

    alias_method :orig_request, :request
    def request(*args)
      if @mock_request
        @mock_client.request(*args)
      else
        orig_request(*args)
      end
    end

    private

    def _request(request_opts)
      mock_client.execute(request_opts)
    end
  end
end

module SharedContext
  extend RSpec::SharedContext
  let(:mock_client) { Payout.mock_client }

  def should_execute_with(opts)
    expect(mock_client).to receive(:execute).with(hash_including(opts)).once do
      # Here the response doesn't matter since we're only testing what the
      # client should receive not what it should respond with.
      double('response', code: 200, body: nil)
    end

    subject
  end

  def should_request_with(*args)
    Payout.mock_request
    Payout.api_token = 'api_token'
    Payout.api_secret = 'api_secret'

    expect(mock_client).to receive(:request).with(*args).once do |*args|
      Payout.orig_request(*args)
    end

    allow(mock_client).to receive(:execute).once do
      double('response', code: 200, body: nil)
    end

    subject
  end
end

RSpec.configure do |config|
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  config.include SharedContext

  config.around(:each) do |example|
    Payout.within_test do
      example.run
    end
  end

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # These two settings work together to allow you to limit a spec run
  # to individual examples or groups you care about by tagging them with
  # `:focus` metadata. When nothing is tagged with `:focus`, all examples
  # get run.
  config.filter_run :focus
  config.run_all_when_everything_filtered = true

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = 'doc'
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end


spec/payout/v1/card_spec.rb
module Payout
  module V1
    RSpec.describe Card do
      let(:token) { 'CTxb2czBhlBrFQSQeabm8bTo' }

      describe '::retrieve' do
        subject { Card.retrieve(token) }
        it { should_request_with(:get, '/v1/cards/%s' % token) }
      end # ::retrieve

      describe '::tokenize' do
        subject { Card.tokenize(params) }

        context 'with required card params' do
          let(:params) do
            {
              name: 'John Smith',
              card_number: '4111111111111111',
              exp_month: '08',
              exp_year: '20'
            }
          end

          it { should_request_with(:post, '/v1/cards', params) }
        end
      end # ::tokenize

      describe '::update_customer' do
        subject { Card.update_customer(params) }

        context 'with required customer params' do
          let(:params) do
            {
              card_token: token,
              name: 'John Smith',
              address_line1: '1 Stockton St.',
              address_line2: '#123',
              address_city: 'San Francisco',
              address_state: 'CA',
              address_zip: '94108'
            }
          end

          it 'should make expected request' do
            should_request_with(:put, "/v1/cards/#{token}/customer",
              params.reject { |x| x == :card_token })
          end
        end
      end # ::update_customer

      describe '::credit' do
        subject { Card.credit(params) }

        context 'with valid params' do
          let(:params) do
            {
              card_token: token,
              amount: 100
            }
          end

          it 'should make expected request' do
            should_request_with(:post, '/v1/cards/%s/credits' % token,
              params.reject { |x| x == :card_token })
          end
        end

        context 'with card_token missing' do
          let(:params) { { amount: 100 } }

          it 'should raise error ArgumentError' do
            expect { subject }.to raise_error ArgumentError,
              'missing card_token'
          end
        end
      end # ::credit
    end # Card
  end # V1
end # Payout


lib/payout/response.rb
require 'json'

##
# Wraps a RestClient response to provide the Payout SDK interface.
module Payout
  class Response
    class << self
      # Handles the response from RestClient, including any errors that may be
      # raised.
      def handle
        new(yield)
      rescue SocketError => e
        _handle_restclient_connection_error(e)
      rescue RestClient::ExceptionWithResponse => e
        _handle_error_with_response(e)
      rescue RestClient::Exception, Errno::ECONNREFUSED => e
        _handle_restclient_connection_error(e)
      end

      private

      def _handle_restclient_connection_error(error)
        case error
        when RestClient::RequestTimeout
          fail TimeoutError
        when RestClient::ServerBrokeConnection
          fail ConnectionError, 'The server broke the connection before the ' \
            'request could complete.'
        when RestClient::SSLCertificateNotVerified
          fail ConnectionError, 'Failed to verify SSL certificate.'
        when SocketError
          fail ConnectionError, 'An unexpected error occurred when trying to ' \
            'connect to Payout. This could be a DNS issue. Check that you can ' \
            'resolve live.payout.com and/or sandbox.payout.com.'
        else
          fail ConnectionError, 'An unexpected error occurred when trying to ' \
            'connect to Payout. Try again or contact us at support@payout.com.'
        end
      end

      def _handle_error_with_response(error)
        if (resp = error.response)
          fail AuthenticationError, 'invalid credentials' if resp.code == 401
          new(resp)
        else
          _handle_restclient_connection_error(error)
        end
      end
    end # Class Methods

    attr_reader :code
    attr_reader :body

    def initialize(response)
      @code = response.code
      @body = response.body
      _parse(response.body)
    end

    def successful?
      [200, 201].include?(code)
    end

    def [](key)
      @_data[key]
    end

    def to_h
      @_data.dup
    end

    private

    def _parse(body)
      @_data = JSON.parse(body || '{}', symbolize_names: true)
    rescue JSON::ParserError
      @_data = {}
    end
  end
end


lib/payout/v1/bank_account.rb
module Payout
  module V1::BankAccount
    class << self
      def retrieve(token)
        Payout.request(:get, "/v1/bank_accounts/#{token}")
      end

      def tokenize(params)
        Payout.request(:post, '/v1/bank_accounts', params)
      end

      def update_customer(params)
        token, params = _extract_token(params.dup)
        Payout.request(:put, "/v1/bank_accounts/#{token}/customer", params)
      end

      def credit(params)
        token, params = _extract_token(params.dup)
        Payout.request(:post, "/v1/bank_accounts/#{token}/credits", params)
      end

      private

      def _extract_token(params)
        token = params.delete(:bank_account_token) or
          fail ArgumentError, 'missing bank_account_token'

        [token, params]
      end
    end
  end # V1::BankAccount
end # Payout


lib/payout.rb
require 'payout/version'
require 'json'
require 'base64'
require 'openssl'
require 'rest-client'

module Payout
  autoload(:Response, 'payout/response')
  autoload(:V1,       'payout/v1')

  class Error < StandardError; end
  class AuthenticationError < Error; end
  class VersionError < Error; end
  class ConnectionError < Error; end
  class TimeoutError < ConnectionError; end

  DEFAULT_API_VERSION = 1
  DEFAULT_OPEN_TIMEOUT = 30
  DEFAULT_READ_TIMEOUT = 80
  SSL_CA_FILE = File.expand_path('../../data/ca-file.crt', __FILE__).freeze

  class << self
    def version
      VERSION
    end

    def api_version
      if const_defined?(:API_VERSION)
        API_VERSION
      else
        self.api_version = DEFAULT_API_VERSION
      end
    end

    def api_url
      @api_url || 'https://live.payout.com'
    end

    def api_token
      @api_token || fail(AuthenticationError,
        'Payout.api_token must be defined')
    end

    def api_secret
      @api_secret || fail(AuthenticationError,
        'Payout.api_secret must be defined')
    end

    def open_timeout
      @open_timeout || DEFAULT_OPEN_TIMEOUT
    end

    def read_timeout
      @read_timeout || DEFAULT_READ_TIMEOUT
    end

    def api_version=(version)
      fail ArgumentError, 'must be an integer' unless version.is_a?(Integer)
      fail ArgumentError, 'must be a positive integer' unless version > 0

      _include_version(version)
    end

    def api_url=(api_url)
      if api_url
        api_url = api_url.dup
        api_url = api_url[0..-2] if api_url[-1] == '/'
      end

      @api_url = api_url.freeze
    end

    def api_token=(token)
      @api_token = token.dup.freeze
    end

    def api_secret=(secret)
      @api_secret = secret.dup.freeze
    end

    def open_timeout=(timeout)
      fail ArgumentError, 'must be an integer' unless timeout.is_a?(Integer)
      @open_timeout = timeout
    end

    def read_timeout=(timeout)
      fail ArgumentError, 'must be an integer' unless timeout.is_a?(Integer)
      @read_timeout = timeout
    end

    def request(verb, path, params = {})
      Response.handle { request!(verb, path, params) }
    end

    def request!(verb, path, params)
      url = _build_request_url(path)
      headers = _default_headers
      body = nil

      case verb
      when :get, :delete
        headers[:params] = params if params && params.any?
      when :post, :put
        headers[:content_type] = 'application/json'
        body = params.to_json if params && params.any?
      else
        fail ArgumentError, "invalid request verb: #{verb.inspect}"
      end

      _request(
        verify_ssl:   OpenSSL::SSL::VERIFY_PEER,
        ssl_ca_file:  SSL_CA_FILE,
        open_timeout: open_timeout,
        read_timeout: read_timeout,
        method:       verb,
        url:          url,
        headers:      headers,
        payload:      body
      )
    end

    private

    def const_missing(name)
      super if const_defined?(:API_VERSION)
      self.api_version = DEFAULT_API_VERSION
      super unless const_defined?(name)
      const_get(name)
    end

    def _include_version(version)
      if const_defined?(:API_VERSION)
        fail VersionError, 'cannot change version after it has been initialized'
      end

      unless (version_const = "V#{version}") && const_defined?(version_const)
        fail ArgumentError, 'unsupported version'
      end

      include const_get(version_const)

      const_set(:API_VERSION, version)
    end

    def _build_request_url(path)
      api_url + (path[0] == '/' ? path : "/#{path}")
    end

    def _default_headers
      {
        authorization: "Basic #{_encoded_credentials}"
      }
    end

    def _encoded_credentials
      Base64.strict_encode64("#{api_token}:#{api_secret}")
    end

    def _request(request_opts)
      RestClient::Request.execute(request_opts)
    end
  end # Class Methods
end # Payout
